Sources/ApodiniAuthorization/Delegates/Authenticator.swift

```diff
@@ -0,0 +1,130 @@
+        self.requirements = requirements
+        self.skipRequirementsForAuthorized = configuration.skipRequirementsForAuthorized
+        self.delegate = Delegate(handler, .required)
+    }
+
+    func handle() throws -> H.Response {
+        if authenticatable.isAuthorized {
+            if skipRequirementsForAuthorized {
+                return try delegate().handle()
+            }
+
+            return try evaluateRequirements(against: try authenticatable())
+        }
+
+        let authenticationScheme = try self.scheme()
+
+        do {
+            return try authenticationAndAuthorize(scheme: authenticationScheme)
+        } catch {
+            // Map any error containing a `AuthorizationErrorReason` option using `mapFailedAuthorization`
+            guard error.apodiniError.option(for: .authorizationErrorReason) != nil else {
+                throw error
+            }
+
+            logger.trace("Authorization on Handler \(H.self) failed with \(error)")
+
+            throw authenticationScheme.mapFailedAuthorization(failedWith: error.apodiniError)
+        }
+    }
+
+    func authenticationAndAuthorize(scheme: Configuration.Scheme) throws -> H.Response {
+        let maybeAuthenticationInfo: Configuration.Scheme.AuthenticationInfo?
+        do {
+            maybeAuthenticationInfo = try scheme.deriveAuthenticationInfo()
+        } catch {
+            throw error.apodiniError(options: .authorizationErrorReason(.invalidAuthenticationRequest))
+        }
+
+        guard let authenticationInfo = maybeAuthenticationInfo else {
+            switch type {
+            case .required:
+                throw authenticationRequired
+            case .optional:
+                // found no authentication information and it isn't required.
+                return try authenticatable.store(into: delegate)()
+                    .handle()
+            }
+        }
+
+        let verifier = try self.verifier()
+
+        let instance: Configuration.Authenticatable
+        do {
+            instance = try verifier.initializeAndVerify(for: authenticationInfo)
+        } catch {
+            throw error.apodiniError(options: .authorizationErrorReason(.failedAuthentication))
+        }
+
+        return try evaluateRequirements(against: instance)
+    }
+
+    func evaluateRequirements(against instance: Configuration.Authenticatable) throws -> H.Response {
+        let result: RequirementResult
+        do {
+            result = try requirements.evaluate(for: instance)
+        } catch {
+            throw error.apodiniError(options: .authorizationErrorReason(.failedAuthorization))
+        }
+
+        switch result {
+        case let .fulfilled(cause), let .undecided(cause): // undecided is a acceptance state as well!
+            logger.trace("Authorization on Handler \(H.self) succeeded with \(cause())")
+            if authenticatable.isAuthorized {
+                return try delegate().handle()
```

Comment:
The `evaluateRequirements(...)` method is not only called once an instance is fully authorized, but also at the beginning of `handle()` if the authenticator finds that there is already an authorized instance (of the same `Authenticatable` type).
See:
https://github.com/Apodini/Apodini/blob/37de2a6567d6e3190764c9d397728409ed9faef3/Sources/ApodiniAuthorization/Delegates/Authenticator.swift#L38-L45

If an authenticator finds that there is already an authorized instance, it skips its own authentication and authorization operations. Only if not turned off via `skipRequirementsForAuthorized` (which is false by default), it will evaluate the authorization requirements against the existing instance.
And that's the case what you commented above. If there is already an existing instance, and the requirements are fulfilled, there is no reason to change something in the environment, because after all nothing changed.